### ISSUE TITLE
Optional logging of validation loss (and other metrics) in KerasModel

### ISSUE THREAD

**Issue by VIGS25** [deepchem/deepchem#1628]
KerasModel (and TensorGraph) currently don't support periodic logging of validation loss.
Would it be a good idea to have this in the `fit_generator` and `fit` API?

This would be an optional argument with the modified API looking something like:

```
def fit(self, dataset, nb_epoch=10, max_checkpoints_to_keep=5, 
      checkpoint_interval=1000, deterministic=False, restore=False, 
      submodel=None, val_dataset=None, eval_interval=1000, **kwargs):
```

It would make it easier to do things like early stopping, which was done in the ChemNet Transfer Learning paper, and in general can be useful as well

**Comment by rbharath** [comment]
This seems like a useful feature. What do you think @peastman?

**Comment by peastman** [comment]
Agreed, this would be very useful.  We should give some thought to how this should work.  For `fit()`, it could work to just provide another dataset to use for validation.  For `fit_generator()` that isn't necessarily possible.  After all, one of the purposes of that method is to support models that take multiple inputs and therefore require more than just the `X` array from a dataset.

You also mentioned the possibility of tracking other metrics, which would also be useful.  So perhaps we can have a unified mechanism that supports all of those things.

**Comment by VIGS25** [comment]
Copying over content from #1633: 

> What do you think the main issues would be?
From what I see, the ops would be the same for the datasets, except for gradients computed on the train dataset. Apart from that, I have had issues wrapping the `train_op` and `loss_tensor` into a clean API, but that can be fixed.

**Comment by peastman** [comment]
I've been thinking about ways to support this.  There are lots of things someone might want to do.  They might want to compute the average loss over the validation set.  Or some other metric over the validation set.  Or they might have some other way of evaluating the model that doesn't specifically involve a validation set.  Once they've done that there are a lot of things they might want to do.  Examples include:

- Write the evaluation metric to the console.
- Log it to Tensorboard.
- End fitting (early stopping).
- Adjust the learning rate.
- Save a copy of the model parameters to disk.

It would be good to create a mechanism that can support all of these and others.  I'm thinking along the lines of some sort of callback mechanism.  The code might look something like this:

```python
callback = ValidationCallback(validation_dataset, interval=10000)
model.fit(..., callback=callback)
```

We could provide a variety of standard callbacks, and users could also write their own.

Note that Keras also supports a callback mechanism for model fitting, but I don't like some parts of how it's designed.  I think we can do better.

**Comment by VIGS25** [comment]
So the current PR does the following things:

- Write the evaluation metric to TensorBoard
- Log it to TensorBoard
- Compute average loss over the validation set
- Compute metrics on the validation set, if metrics are supplied
- Save a copy of model parameters to disk

But a more general setup would definitely be great!

**Comment by peastman** [comment]
I'm getting ready to start on this.  Here's the API I'm thinking of.

You can pass an arbitrary list of callbacks to any of the fitting methods.  A callback can be any callable object.  It gets invoked after each step, with the KerasModel being passed as the only argument.  (Possibly the index of the current step will also be passed.  I'm still thinking about that.)  A callback can raise an exception to signal the model to end fitting.

There will be a class called ValidationCallback to handle validation.  When you create it you'll specify a Dataset and an interval at which to perform validation (measured in training steps).  It will only support models that can be directly applied to a Dataset.  For other models, you need to write your own callback.  You can optionally specify one or more Metrics which will be logged along with the loss.

I'm still thinking about how best to handle two other features: early stopping, and saving a copy of the best model parameters so far to a particular directory.  You might potentially want to do all of these things at once, and you don't want several callbacks to each independently repeat the validation.  Should all these be features of a single class?  Or do we allow dependencies between callbacks, so one of them computes the loss and then others make decisions based on it?  Perhaps one callback could wrap another one?  I'm still thinking about it.

**Comment by VIGS25** [comment]
So from what I understand, there is a class (BaseLogger), would be invoked at every step and handle loss computation at every step, and logs the training loss periodically. There is a ValidationCallBack, that is setup if `validation_dataset` is not None, and that handles metrics and losses for validation dataset. Is this correct?

I think saving a copy of the best model parameters to directory can be a property of the BaseLogger.
Can't we have EarlyStopping inherit from ValidationCallBack? - It does most things similar to ValidationCallBack, but just has an additional stopping criteria.

**Comment by peastman** [comment]
What is BaseLogger?  That wasn't part of my proposal.  Is that something you're suggesting?  If so, could you describe in more detail what it would do?

> Can't we have EarlyStopping inherit from ValidationCallBack?

I considered that, but what if you want both early stopping (when the validation score stops improving) and also saving a copy of the model whenever you get a new best validation score?  If those were separate classes, and each of them extended ValidationCallback, that would mean running the validation twice.

**Comment by VIGS25** [comment]
I borrowed the `BaseLogger` term from Keras Callbacks, but it was motivated by your invoked "after every step" statement: 
> A callback can be any callable object. It gets invoked after each step, with the KerasModel being passed as the only argument.

Using the same idea as Keras, this CallBack would compute the loss at every step, and then display every certain iterations. This is used by default in every model. It can slso compute some metrics on the training set, if needed. 

In case of Keras (https://github.com/keras-team/keras/blob/master/keras/callbacks.py), the EarlyStoppingCallback keeps track of the best weights if a toggle is turned on. So having one class should be sufficient.
